File: android/src/main/java/expo/modules/mpvplayer/MpvPlayerView.kt
package expo.modules.mpvplayer

import android.app.Activity
import android.app.Application
import android.content.Context
import android.content.ContextWrapper
import android.graphics.Color
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.SurfaceHolder
import android.view.SurfaceView
import android.view.ViewGroup
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import expo.modules.mpvplayer.nativeplayer.engine.PlayerEngine
import expo.modules.mpvplayer.nativeplayer.engine.VideoLoadConfig

class MpvPlayerView(context: Context, appContext: AppContext) : ExpoView(context, appContext),
    PlayerEngine.Delegate, SurfaceHolder.Callback {

    companion object {
        private const val TAG = "MpvPlayerView"
        private const val RESUME_RECOVERY_DELAY_MS = 300L
    }

    val onLoad by EventDispatcher()
    val onPlaybackStateChange by EventDispatcher()
    val onProgress by EventDispatcher()
    val onError by EventDispatcher()
    val onTracksReady by EventDispatcher()
    val onPictureInPictureChange by EventDispatcher()
    val onEnd by EventDispatcher()

    private val surfaceView = SurfaceView(context)
    private var renderer: MPVLayerRenderer? = null
    private var rendererVoDriver: String? = null
    private val pipController = PiPController(context, appContext)
    private var surfaceReady = false
    private var rendererStarted = false
    private var pendingConfig: VideoLoadConfig? = null
    private var currentUrl: String? = null
    private var currentLoop = false
    private var cachedPosition = 0.0
    private var cachedDuration = 0.0
    private var zoomed = false
    private var intendedPlayState = false
    private val uiHandler = Handler(Looper.getMainLooper())
    private var hostActivity: Activity? = null
    private var lifecycleCallbacks: Application.ActivityLifecycleCallbacks? = null
    private var lifecycleRegistered = false
    private val recoverResumeRunnable = Runnable { runResumeRecovery() }

    init {
        setBackgroundColor(Color.BLACK)
        surfaceView.layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT,
        )
        surfaceView.holder.addCallback(this)
        surfaceView.addOnLayoutChangeListener { _, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom ->
            val w = right - left
            val h = bottom - top
            val oldW = oldRight - oldLeft
            val oldH = oldBottom - oldTop
            if (w > 0 && h > 0 && (w != oldW || h != oldH)) renderer?.updateSurfaceSize(w, h)
        }
        addView(surfaceView)
        pipController.setPlayerView(surfaceView)
        pipController.delegate = object : PiPController.Delegate {
            override fun onPlay() = play()
            override fun onPause() = pause()
            override fun onSeekBy(seconds: Double) = seekBy(seconds)
            override fun onPictureInPictureModeChanged(isInPiP: Boolean) {
                uiHandler.removeCallbacksAndMessages(null)
                uiHandler.postDelayed({ syncSurfaceSizeToView() }, 100)
                if (isInPiP) uiHandler.postDelayed({ syncSurfaceSizeToView() }, 500)
                onPictureInPictureChange(mapOf("isActive" to isInPiP))
            }
        }
        registerLifecycleCallbacks()
    }

    private fun ensureRendererStarted(voDriver: String?) {
        if (rendererStarted) { loadPending(); return }
        val wanted = voDriver ?: "gpu-next"
        if (renderer == null || rendererVoDriver != wanted) {
            renderer?.stop()
            renderer = MPVLayerRenderer(context, wanted)
            rendererVoDriver = wanted
        }
        renderer?.delegate = this
        renderer?.start(PlayerEngine.Owner.EMBEDDED_VIEW) {
            rendererStarted = true
            renderer?.playbackResumeIntent = intendedPlayState
            surfaceView.holder.surface?.takeIf { it.isValid }?.let { renderer?.attachSurface(it) }
            syncSurfaceSizeToView()
            loadPending()
        }
    }

    private fun loadPending() {
        if (!rendererStarted || !surfaceReady) return
        pendingConfig?.let { config ->
            currentUrl = config.url
            currentLoop = config.loop
            renderer?.load(config)
            if (config.autoplay) play()
            onLoad(mapOf("url" to config.url))
        }
        pendingConfig = null
    }

    override fun surfaceCreated(holder: SurfaceHolder) {
        surfaceReady = true
        if (rendererStarted) {
            renderer?.attachSurface(holder.surface)
            syncSurfaceSizeToView()
        }
        pendingConfig?.let { ensureRendererStarted(it.voDriver) }
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
        if (width > 0 && height > 0) renderer?.updateSurfaceSize(width, height)
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
        surfaceReady = false
        renderer?.detachSurface()
    }

    fun loadVideo(config: VideoLoadConfig) {
        if (currentUrl == config.url && currentLoop == config.loop) return
        pendingConfig = config
        if (surfaceReady) ensureRendererStarted(config.voDriver)
    }

    fun play() {
        intendedPlayState = true
        renderer?.playbackResumeIntent = true
        renderer?.play()
        pipController.setPlaybackRate(1.0)
    }
    fun pause() {
        intendedPlayState = false
        renderer?.playbackResumeIntent = false
        renderer?.pause()
        pipController.setPlaybackRate(0.0)
    }
    fun destroy() {
        renderer?.stop()
        rendererStarted = false
        pendingConfig = null
        currentUrl = null
        intendedPlayState = false
    }
    fun seekTo(position: Double) = renderer?.seekTo(position) ?: Unit
    fun seekBy(offset: Double) = renderer?.seekBy(offset) ?: Unit
    fun setSpeed(speed: Double) = renderer?.setSpeed(speed) ?: Unit
    fun setMute(muted: Boolean) = renderer?.setMute(muted) ?: Unit
    fun getSpeed(): Double = renderer?.getSpeed() ?: 1.0
    fun isPaused(): Boolean = renderer?.isPausedState ?: true
    fun getCurrentPosition(): Double = cachedPosition
    fun getDuration(): Double = cachedDuration

    fun startPictureInPicture() = pipController.startPictureInPicture()
    fun stopPictureInPicture() = pipController.stopPictureInPicture()
    fun isPictureInPictureSupported(): Boolean = pipController.isPictureInPictureSupported()
    fun isPictureInPictureActive(): Boolean = pipController.isPictureInPictureActive()

    fun getSubtitleTracks(): List<Map<String, Any>> = renderer?.getSubtitleTracks() ?: emptyList()
    fun setSubtitleTrack(trackId: Int) = renderer?.setSubtitleTrack(trackId) ?: Unit
    fun disableSubtitles() = renderer?.disableSubtitles() ?: Unit
    fun getCurrentSubtitleTrack(): Int = renderer?.getCurrentSubtitleTrack() ?: 0
    fun addSubtitleFile(url: String, select: Boolean = true) = renderer?.addSubtitleFile(url, select) ?: Unit
    fun setSubtitlePosition(position: Int) = renderer?.setSubtitlePosition(position) ?: Unit
    fun setSubtitleScale(scale: Double) = renderer?.setSubtitleScale(scale) ?: Unit
    fun setSubtitleDelay(seconds: Double) = renderer?.setSubtitleDelay(seconds) ?: Unit
    fun setSubtitleMarginY(margin: Int) = renderer?.setSubtitleMarginY(margin) ?: Unit
    fun setSubtitleAlignX(alignment: String) = renderer?.setSubtitleAlignX(alignment) ?: Unit
    fun setSubtitleAlignY(alignment: String) = renderer?.setSubtitleAlignY(alignment) ?: Unit
    fun setSubtitleStyle(config: Map<String, Any>) = renderer?.setSubtitleStyle(config) ?: Unit
    fun setSubtitleFontSize(size: Int) = renderer?.setSubtitleFontSize(size) ?: Unit
    fun setSubtitleBorderStyle(style: String) = renderer?.setSubtitleBorderStyle(style) ?: Unit
    fun setSubtitleBackgroundColor(color: String) = renderer?.setSubtitleBackgroundColor(color) ?: Unit
    fun setSubtitleAssOverride(mode: String) = renderer?.setSubtitleAssOverride(mode) ?: Unit
    fun getAudioTracks(): List<Map<String, Any>> = renderer?.getAudioTracks() ?: emptyList()
    fun setAudioTrack(trackId: Int) = renderer?.setAudioTrack(trackId) ?: Unit
    fun getCurrentAudioTrack(): Int = renderer?.getCurrentAudioTrack() ?: 0
    fun setAudioDelay(seconds: Double) = renderer?.setAudioDelay(seconds) ?: Unit
    fun setVolumeBoost(percent: Int) = renderer?.setVolumeBoost(percent) ?: Unit
    fun setDialogueBoost(enabled: Boolean) = renderer?.setDialogueBoost(enabled) ?: Unit
    fun setMonoDownmix(enabled: Boolean) = renderer?.setMonoDownmix(enabled) ?: Unit
    fun setZoomedToFill(value: Boolean) { zoomed = value; renderer?.setZoomedToFill(value) }
    fun isZoomedToFill(): Boolean = zoomed
    fun getTechnicalInfo(): Map<String, Any> = renderer?.getTechnicalInfo() ?: emptyMap()

    override fun onPositionChanged(position: Double, duration: Double, cacheSeconds: Double) {
        cachedPosition = position; cachedDuration = duration
        if (pipController.isPictureInPictureActive()) pipController.setCurrentTime(position, duration)
        onProgress(mapOf("position" to position, "duration" to duration, "progress" to if (duration > 0) position / duration else 0.0, "cacheSeconds" to cacheSeconds))
    }
    override fun onPauseChanged(isPaused: Boolean) {
        pipController.setPlaybackRate(if (isPaused) 0.0 else 1.0)
        onPlaybackStateChange(mapOf("isPaused" to isPaused, "isPlaying" to !isPaused))
    }
    override fun onLoadingChanged(isLoading: Boolean) = onPlaybackStateChange(mapOf("isLoading" to isLoading))
    override fun onReadyToSeek() = onPlaybackStateChange(mapOf("isReadyToSeek" to true))
    override fun onTracksReady() = onTracksReady(emptyMap<String, Any>())
    override fun onError(message: String) = onError(mapOf("error" to message))
    override fun onVideoDimensionsChanged(width: Int, height: Int) = pipController.setVideoDimensions(width, height)
    override fun onPlaybackEnded() = onEnd(emptyMap<String, Any>())

    private fun syncSurfaceSizeToView() {
        if (!surfaceReady) return
        val w = surfaceView.width
        val h = surfaceView.height
        if (w > 0 && h > 0) renderer?.updateSurfaceSize(w, h)
    }

    private fun runResumeRecovery() {
        if (!rendererStarted || currentUrl == null) return
        if (pipController.isPictureInPictureActive()) return
        if (intendedPlayState) return
        val surface = surfaceView.holder.surface?.takeIf { it.isValid }
        Log.i(TAG, "Resume recovery: paused, surfaceValid=${surface != null}")
        renderer?.recoverVideoOutput(surface)
    }

    private fun registerLifecycleCallbacks() {
        if (lifecycleRegistered || !DeviceKind.isTelevision(context)) return
        val app = context.applicationContext as? Application ?: return
        lifecycleCallbacks = object : Application.ActivityLifecycleCallbacks {
            override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}
            override fun onActivityStarted(activity: Activity) {}
            override fun onActivityResumed(activity: Activity) {
                val host = hostActivity ?: findActivity().also { hostActivity = it }
                if (activity !== host || currentUrl == null || intendedPlayState || pipController.isPictureInPictureActive()) return
                uiHandler.removeCallbacks(recoverResumeRunnable)
                uiHandler.postDelayed(recoverResumeRunnable, RESUME_RECOVERY_DELAY_MS)
            }
            override fun onActivityPaused(activity: Activity) {}
            override fun onActivityStopped(activity: Activity) {}
            override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
            override fun onActivityDestroyed(activity: Activity) {}
        }
        app.registerActivityLifecycleCallbacks(lifecycleCallbacks)
        lifecycleRegistered = true
    }

    private fun unregisterLifecycleCallbacks() {
        uiHandler.removeCallbacks(recoverResumeRunnable)
        if (!lifecycleRegistered) return
        val app = context.applicationContext as? Application
        lifecycleCallbacks?.let { app?.unregisterActivityLifecycleCallbacks(it) }
        lifecycleCallbacks = null
        lifecycleRegistered = false
    }

    private fun findActivity(): Activity? {
        appContext.currentActivity?.let { return it }
        var ctx: Context = context
        while (ctx is ContextWrapper) {
            if (ctx is Activity) return ctx
            ctx = ctx.baseContext
        }
        return null
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        uiHandler.removeCallbacksAndMessages(null)
        unregisterLifecycleCallbacks()
        pipController.stopPictureInPicture()
        renderer?.stop()
        renderer?.delegate = null
    }
}

File: ios/MpvPlayerView.swift
import AVFoundation
import ExpoModulesCore
import UIKit

struct VideoLoadConfig {
    let url: URL
    var headers: [String: String]?
    var externalSubtitles: [String]?
    var startPosition: Double?
    var autoplay: Bool
    var initialSubtitleId: Int?
    var initialAudioId: Int?
    var loop: Bool
    var cacheEnabled: String?
    var cacheSeconds: Int?
    var demuxerMaxBytes: Int?
    var demuxerMaxBackBytes: Int?
    var cachePause: Bool?
    var cachePauseInitial: Bool?
    var cachePauseWaitSeconds: Double?
}

class MpvPlayerView: ExpoView {
    private let engine = MPVPlayerEngine()
    private let videoContainer = UIView()
    let onLoad = EventDispatcher()
    let onPlaybackStateChange = EventDispatcher()
    let onProgress = EventDispatcher()
    let onError = EventDispatcher()
    let onTracksReady = EventDispatcher()
    let onPictureInPictureChange = EventDispatcher()
    let onEnd = EventDispatcher()

    required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)
        clipsToBounds = true
        backgroundColor = .black
        videoContainer.translatesAutoresizingMaskIntoConstraints = false
        videoContainer.backgroundColor = .black
        videoContainer.clipsToBounds = true
        addSubview(videoContainer)
        NSLayoutConstraint.activate([
            videoContainer.topAnchor.constraint(equalTo: topAnchor), videoContainer.bottomAnchor.constraint(equalTo: bottomAnchor),
            videoContainer.leadingAnchor.constraint(equalTo: leadingAnchor), videoContainer.trailingAnchor.constraint(equalTo: trailingAnchor),
        ])
        engine.delegate = self
        videoContainer.layer.addSublayer(engine.displayLayer)
        do { try engine.start() } catch { onError(["error": "Failed to start renderer: \(error.localizedDescription)"]) }
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        CATransaction.begin(); CATransaction.setDisableActions(true)
        engine.displayLayer.frame = videoContainer.bounds
        engine.displayLayer.contentsScale = contentScaleFactor
        CATransaction.commit()
    }

    func loadVideo(config: VideoLoadConfig) { engine.loadVideo(config: config) }
    func setNowPlayingMetadata(_ metadata: [String: String], artworkHeaders: [String: String]? = nil) { engine.setNowPlayingMetadata(metadata, artworkHeaders: artworkHeaders) }
    func play() { engine.play() }
    func pause() { engine.pause() }
    func destroy() { engine.destroy() }
    func seekTo(position: Double) { engine.seekTo(position: position) }
    func seekBy(offset: Double) { engine.seekBy(offset: offset) }
    func setSpeed(speed: Double) { engine.setSpeed(speed: speed) }
    func setMute(muted: Bool) { engine.setMute(muted) }
    func getSpeed() -> Double { engine.getSpeed() }
    func isPaused() -> Bool { engine.isPaused() }
    func getCurrentPosition() -> Double { engine.getCurrentPosition() }
    func getDuration() -> Double { engine.getDuration() }
    func startPictureInPicture() { engine.startPictureInPicture() }
    func stopPictureInPicture() { engine.stopPictureInPicture() }
    func isPictureInPictureSupported() -> Bool { engine.isPictureInPictureSupported() }
    func isPictureInPictureActive() -> Bool { engine.isPictureInPictureActive() }
    func getSubtitleTracks(completion: @escaping ([[String: Any]]) -> Void) { engine.getSubtitleTracks(completion: completion) }
    func setSubtitleTrack(_ id: Int) { engine.setSubtitleTrack(id) }
    func disableSubtitles() { engine.disableSubtitles() }
    func getCurrentSubtitleTrack(completion: @escaping (Int) -> Void) { engine.getCurrentSubtitleTrack(completion: completion) }
    func addSubtitleFile(url: String, select: Bool = true) { engine.addSubtitleFile(url: url, select: select) }
    func getAudioTracks(completion: @escaping ([[String: Any]]) -> Void) { engine.getAudioTracks(completion: completion) }
    func setAudioTrack(_ id: Int) { engine.setAudioTrack(id) }
    func getCurrentAudioTrack(completion: @escaping (Int) -> Void) { engine.getCurrentAudioTrack(completion: completion) }
    func setAudioDelay(_ seconds: Double) { engine.setAudioDelay(seconds) }
    func setVolumeBoost(_ percent: Int) { engine.setVolumeBoost(percent) }
    func setDialogueBoost(_ enabled: Bool) { engine.setDialogueBoost(enabled) }
    func setMonoDownmix(_ enabled: Bool) { engine.setMonoDownmix(enabled) }
    func setSubtitlePosition(_ v: Int) { engine.setSubtitlePosition(v) }
    func setSubtitleScale(_ v: Double) { engine.setSubtitleScale(v) }
    func setSubtitleDelay(_ v: Double) { engine.setSubtitleDelay(v) }
    func setSubtitleMarginY(_ v: Int) { engine.setSubtitleMarginY(v) }
    func setSubtitleAlignX(_ v: String) { engine.setSubtitleAlignX(v) }
    func setSubtitleAlignY(_ v: String) { engine.setSubtitleAlignY(v) }
    func setSubtitleStyle(config: [String: Any]) { engine.setSubtitleStyle(config: config) }
    func setSubtitleFontSize(_ v: Int) { engine.setSubtitleFontSize(v) }
    func setSubtitleBackgroundColor(_ v: String) { engine.setSubtitleBackgroundColor(v) }
    func setSubtitleBorderStyle(_ v: String) { engine.setSubtitleBorderStyle(v) }
    func setSubtitleAssOverride(_ v: String) { engine.setSubtitleAssOverride(v) }
    func setZoomedToFill(_ v: Bool) { engine.setZoomedToFill(v) }
    func isZoomedToFill() -> Bool { engine.isZoomedToFill() }
    func getTechnicalInfo(completion: @escaping ([String: Any]) -> Void) { engine.getTechnicalInfo(completion: completion) }

    deinit { engine.shutdown() }
}

extension MpvPlayerView: MPVPlayerEngineDelegate {
    func engine(_ engine: MPVPlayerEngine, didLoad url: URL) { onLoad(["url": url.absoluteString]) }
    func engine(_ engine: MPVPlayerEngine, didUpdateProgress position: Double, duration: Double, cacheSeconds: Double) {
        onProgress(["position": position, "duration": duration, "progress": duration > 0 ? position / duration : 0, "cacheSeconds": cacheSeconds])
    }
    func engine(_ engine: MPVPlayerEngine, didChangePause isPaused: Bool) { onPlaybackStateChange(["isPaused": isPaused, "isPlaying": !isPaused]) }
    func engine(_ engine: MPVPlayerEngine, didChangeLoading isLoading: Bool) { onPlaybackStateChange(["isLoading": isLoading]) }
    func engine(_ engine: MPVPlayerEngine, didBecomeReadyToSeek ready: Bool) { onPlaybackStateChange(["isReadyToSeek": ready]) }
    func engine(_ engine: MPVPlayerEngine, didBecomeTracksReady ready: Bool) { onTracksReady([:]) }
    func engine(_ engine: MPVPlayerEngine, didChangePictureInPicture isActive: Bool) { onPictureInPictureChange(["isActive": isActive]) }
    func engine(_ engine: MPVPlayerEngine, didFailWithError message: String) { onError(["error": message]) }
    func engine(_ engine: MPVPlayerEngine, didDetectHDRMode mode: HDRMode, fps: Double) {
        #if os(tvOS)
        setDisplayCriteria(for: mode, fps: Float(fps))
        #endif
    }
    func engineDidReachEnd(_ engine: MPVPlayerEngine) { onEnd([:]) }
}

#if os(tvOS)
import AVKit
import CoreMedia

extension MpvPlayerView {
    private func makeHDRFormat(_ mode: HDRMode) -> CMFormatDescription? {
        if case .sdr = mode { return nil }
        var extensions: [String: Any] = [kCMFormatDescriptionExtension_FullRangeVideo as String: true]
        extensions[kCMFormatDescriptionExtension_ColorPrimaries as String] = kCMFormatDescriptionColorPrimaries_ITU_R_2020
        if case .hlg = mode {
            extensions[kCMFormatDescriptionExtension_TransferFunction as String] = kCMFormatDescriptionTransferFunction_ITU_R_2100_HLG
        } else {
            extensions[kCMFormatDescriptionExtension_TransferFunction as String] = kCMFormatDescriptionTransferFunction_SMPTE_ST_2084_PQ
        }
        extensions[kCMFormatDescriptionExtension_YCbCrMatrix as String] = kCMFormatDescriptionYCbCrMatrix_ITU_R_2020
        var out: CMFormatDescription?
        let status = CMVideoFormatDescriptionCreate(allocator: kCFAllocatorDefault, codecType: kCMVideoCodecType_HEVC, width: 3840, height: 2160, extensions: extensions as CFDictionary, formatDescriptionOut: &out)
        return status == noErr ? out : nil
    }
    private func setDisplayCriteria(for mode: HDRMode, fps: Float) {
        guard #available(tvOS 17.0, *), let window else { return }
        guard let format = makeHDRFormat(mode) else { window.avDisplayManager.preferredDisplayCriteria = nil; return }
        window.avDisplayManager.preferredDisplayCriteria = AVDisplayCriteria(refreshRate: fps, formatDescription: format)
    }
}
#endif
